[Sol -> Yul] Checked signed arithmetic and modulo.

ekpyron opened the pull request.

ekpyron committed:
Implement signed addition for sol->yul code generation.

ekpyron committed:
Implement signed subtraction for sol->yul code generation.

ekpyron committed:
Unify style of checked integer division for sol->yul code generation.

ekpyron committed:
Implement signed multiplication for sol->yul code generation.

ekpyron commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -321,63 +321,59 @@ string YulUtilFunctions::roundUpFunction()
 			function <functionName>(x, y) -> product {
-				if and(iszero(iszero(x)), lt(div(<mask>, x), y)) { revert(0, 0) }
-				<?shortType>
-					product := mulmod(x, y, <powerOfTwo>)
```
Not sure why I used a ``mulmod`` here earlier - that's the main advantage in detecting overflows *before* the operation - if the checks pass it *won't* overflow, so we can just simply multiply here. That's the reason for removing the comment above as well, since it'd only work with ``mulmod`` here, but that's a good bit more expensive, so I'd take this as a good reason to disregard the option mentioned in the comment.

ekpyron committed:
Implement checked mod for sol->yul code generation.

Marenz commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -379,27 +379,26 @@ string YulUtilFunctions::roundUpFunction()
 		return
 			Whiskers(R"(
 			function <functionName>(x, y) -> sum {
-				<?shortType>
```
So the `shortType` stuff wasn't really needed?

ekpyron replied:
It was needed in the old version (that checked for overflow *after* adding), but is no longer *needed* for the new version (that checks for overflow *before* adding).

For short types the  new version is actually cheaper - but I'm just realizing that for ``uint256`` the old version was in fact slightly cheaper, so maybe it makes sense to keep the special case for that...

ekpyron replied:
The ``mul`` case is different - there it was entirely superfluous...

ekpyron replied:
But yeah - I think I'll change the unsigned non-short-type case back to the old version (reintroducing the case distinction) - adding two ``uint256``s is probably among the most common operations, so I guess it makes sense to sacrifice some brevity here for cheaper runtime costs...
Although... if we add a new optimizer rule that changes ``sub(2^256-1, x)`` to ``not(x)`` (which we should probably have anyways, if it's in fact correct), then it'd only be 3 gas difference... I'll think about it tomorrow...

ekpyron replied:
Coolest thing would be, if the optimizer recognized, that it can move the check across the ``add`` and save 3 gas by it :-D. But that's probably utopian :-).

ekpyron replied:
Added a comment above to reconsider this against real world examples at a later point. https://github.com/ethereum/solidity/pull/6935 will reduce the gas difference to 3 gas and at the moment it's unclear whether making it easier for the optimizer to move up the revert condition is more beneficial (or in fact harmful) in contrast to "pre-optimizing" the 256-bit case here, so for the time being I'd opt for simpler code here.

Marenz commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -379,27 +379,26 @@ string YulUtilFunctions::roundUpFunction()
-					if lt(sum, x) { revert(0, 0) }
-				</shortType>
+				<?signed>
+				if and(iszero(slt(x, 0)), sgt(y, sub(<maxValue>, x))) { revert(0, 0) }
```
I might help if you add a comment above here like 
```
if x >= 0 && y > (maxVal - x) -> OVERFLOW
if x < 0 && y < (minVal - x)  -> OVERFLOW
```

ekpyron replied:
Might even help to distinguish between "overflow" and "underflow".

Marenz commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -458,18 +458,26 @@ string YulUtilFunctions::overflowCheckedIntDivFunction(IntegerType const& _type)
 			function <functionName>(x, y) -> diff {
+				<?signed>
+				if and(iszero(slt(y, 0)), slt(x, add(<minValue>, y))) { revert(0, 0) }
+				if and(slt(y, 0), sgt(x, add(<maxValue>, y))) { revert(0, 0) }
```
And here? At least it helped me..
```
if y >= 0 && x < (minVal + y) => OVERFLOW
if y < 0 && x > (maxVal + y) => OVERFLOW
```

Marenz commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -403,31 +403,30 @@ string YulUtilFunctions::overflowCheckedIntAddFunction(IntegerType const& _type)
+				if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(<minValue>, y))) { revert(0, 0) }
+				if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(<maxValue>, y))) { revert(0, 0) }
+				<!signed>
+				if and(iszero(iszero(x)), lt(div(<maxValue>, x), y)) { revert(0, 0) }
```
`if x != 0 && (maxVal / x) < y`

Marenz commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -403,31 +403,30 @@ string YulUtilFunctions::overflowCheckedIntAddFunction(IntegerType const& _type)
+				if and(and(sgt(x, 0), slt(y, 0)), slt(y, sdiv(<minValue>, x))) { revert(0, 0) }
+				if and(and(slt(x, 0), sgt(y, 0)), slt(x, sdiv(<minValue>, y))) { revert(0, 0) }
+				if and(and(slt(x, 0), slt(y, 0)), slt(x, sdiv(<maxValue>, y))) { revert(0, 0) }
+				<!signed>
```
```
if x > 0 && y > 0 && x > (maxVal / y) 
if x > 0 && y < 0 && y < (minVal / x)
if x < 0 && y > 0 && x < (minVal / y)
if x < 0 && y < 0 && x < (maxVal / y)
```

Marenz reviewed (changes requested):
Apart from my suggested comments, LGTM

ekpyron commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -379,71 +379,78 @@ string YulUtilFunctions::roundUpFunction()
 			function <functionName>(x, y) -> r {
 				if iszero(y) { revert(0, 0) }
 				<?signed>
-				// x / -1 == x
```
I don't quite get what the original comment was supposed to mean ;-).

ekpyron commented:
Added comments as requested. I put them in separate commits (I also changed the argument order of one of the checks to be consistent with the others), but I'll squash it together before merging.

ekpyron commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -379,71 +379,78 @@ string YulUtilFunctions::roundUpFunction()
+				if and(slt(x, 0), slt(y, sub(<minValue>, x))) { revert(0, 0) }
+				<!signed>
+				// overflow, if x > (maxValue - y)
+				if gt(x, sub(<maxValue>, y)) { revert(0, 0) }
```
Note that with #6935 this will reduce to ``gt(x, not(y))`` for 256-bit unsigned integers.

ekpyron commented:
I moved all comments to C++ comments - they shouldn't end up in and clutter generated code anyways and having comments involving ``<`` and ``>`` will confuse Whiskers :-).

ekpyron commented:
Squashed a bit.

ekpyron commented:
Hm... I'm still not entirely happy with the comments - currently it'll add blank lines with spaces in them to the generated IR code... I think I'll still try to solve that...

ekpyron commented:
Or well... we have a lot of that going on anyways due to indenting the closing parentheses of the raw literals... so maybe it's just fine...

ekpyron commented:
Closing and opening the raw strings like I did now violates the coding style a bit as well I guess...

ekpyron commented:
Well, I leave it as it is for now - but I'm open to changing it if someone requests it.

ekpyron commented:
Rebased. On top of this I added correctness proofs of the checks in https://github.com/ethereum/solidity/pull/6944.

Marenz commented on libsolidity/codegen/ir/IRGeneratorForStatements.cpp:
```diff
@@ -1113,6 +1113,10 @@ string IRGeneratorForStatements::binaryOperation(
 				fun = m_utils.overflowCheckedIntDivFunction(*type);
 				break;
+			case Token::Mod:
+				solAssert(!type->isSigned(), "");
```
`mod` doesn't support signed?

ekpyron replied:
Added signed ``mod``, resp. ``smod``.

chriseth commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -379,27 +379,38 @@ string YulUtilFunctions::roundUpFunction()
-				</shortType>
+				<?signed>
+				)"
+				// overflow, if x >= 0 and y > (maxValue - x)
```
Why not keep the comments in the yul code?

Marenz replied:
He wrote / said earlier that the `>`/`<` confused whiskers

Marenz replied:
maybe we need to update the regex it uses

chriseth replied:
ah, makes sense. Maybe we should limit Whiskers template names to `[a-zA-Z0-9$_]` or something.

chriseth commented:
Will rebase this now

chriseth commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -452,25 +477,60 @@ string YulUtilFunctions::overflowCheckedIntDivFunction(IntegerType const& _type)
 			function <functionName>(x, y) -> diff {
+				<?signed>
+				)"
+				// underflow, if y >= 0 and x < (minValue - y)
```
```suggestion
				// underflow, if y >= 0 and x < (minValue + y)
```
(fixing myself, but please check)

chriseth commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -452,25 +477,60 @@ string YulUtilFunctions::overflowCheckedIntDivFunction(IntegerType const& _type)
+				)"
+				// underflow, if y >= 0 and x < (minValue - y)
+				R"(
+				if and(iszero(slt(y, 0)), slt(x, add(<minValue>, y))) { revert(0, 0) }
```
```suggestion
				if and(iszero(slt(y, 0)), slt(x, sub(<maxValue>, y))) { revert(0, 0) }
```
(fixing myself, but please check)
